main.c

```diff
@@ -21,23 +22,72 @@ static void die_on(bool condition, char *string, ...)
     }
 }
 
-struct bzimage_params {
-    size_t bzimage_addr;
+struct boot_params {
+    size_t kernel_addr;
     size_t initrd_addr;
     size_t initrd_size_addr;
     size_t cmdline_addr;
 };
 
-static struct bzimage_params initialize_from_fw_cfg()
+enum boot_protocol {
+    UNKNOWN,
+    LINUX,
+    ELF,
+};
+
+static struct boot_params get_boot_params_from_fw_cfg();
+static enum boot_protocol get_boot_protocol(struct boot_params);
+void linux_boot(struct boot_params);
+void elf_boot(struct boot_params);
+
+int main(void)
+{
+    struct boot_params params = get_boot_params_from_fw_cfg();
+    enum boot_protocol boot_protocol = get_boot_protocol(params);
+
+    switch (boot_protocol) {
+    case LINUX:
+        linux_boot(params);
+        break;
+    case ELF:
+        elf_boot(params);
+        break;
+    default:
+        die_on(true, "Unknown boot method\n");
```

Comment:
User will see this error if they misconfigure things and this is not actionable. What about "Failed to recognize kernel file format. Supported formats are bzImage and 32-bit Multiboot ELF".

Reply:
That's right. I also converted the other user facing error messages.